### DISCUSSION TITLE
Command that moves the focus to the next occupied tag

### DISCUSSION THREAD

**Discussion by gabberdancecat** [leftwm/leftwm#961] [Ideas]
Hello, 
I think it would be really nice to have a command that cycles through occupied tags instead of simply adjacent ones (like ```FocusNextTag``` and ```FocusPreviousTag```). For example, if I'm focused on tag 3 and tags 1,3,7,9 are occupied, then going to the next will shift focus to tag 7 (the next occupied tag). I usually have windows open in very distant tags, like with a browser in tag 7, mail client in tag 5, and a few terminals open in tag 1, so this would be very helpful.

Because such a command like this doesn't exist yet, I created a bash script that does the same action, albeit very scuffed. It first identifies the tag to go to, then goes to that tag by repeating ```leftwm-command "FocusNextTag"``` or ```leftwm-command "FocusPreviousTag"``` as many times as appropriate. I have this bash script run whenever a certain keybind is pressed.

Ideally, I would use ```leftwm-command "GoToTag $x"``` to simply jump to the appropriate tag, but ```GoToTag``` is not available in ```leftwm-command```. So, I have to resort to repeating ```leftwm-command "FocusNextTag"``` or ```leftwm-command "FocusPreviousTag"``` many times in a loop to complete this action. This results in windows flashing as I go from one tag to another, which hurts my eyes, and if I perform this action too frequently, it may even skip a tag or two. 

To make my bash script work better, making ```GoToTag``` an available command in ```leftwm-command``` would solve the issue. But an even better solution would be to implement a whole new command in lefthk that shifts your focus to the next/previous occupied tag (although, this could be a big undertaking).

I wonder how many people would benefit from having a command like this? If possibly many, then creating a new command in lefthk could be beneficial. But if a bash script would suffice, then adding ```GoToTag``` to ```leftwm-command``` and having someone create a better script than mine to perform the action, would be nice.

Here's the bash script I made that cycles through occupied tags (UPDATED, now smooth as butter and should support multi-monitor):
```
#!/bin/bash

##[ Note: Run the command with parameters "+" or "-" to cycle forwards or ]##
##[       backwards, respectively.                                        ]##

# Get info of all tags
FULL=$(leftwm-state -q)

# Keep only info for the current workspace (monitor?)
FULL=$( 
	echo "$FULL" |
	awk -v RS="]}" '/"mine":true[^}]*"focused":true/ { 
		print $0
	}' 
)

# Get the currently focused tag number
NOW=$( 
	echo "$FULL" | 
	awk -v RS='{' '/"focused":true/ { 
		beg = index($0, name);
		print substr($0, beg+8, 1)
	}'
)

# Get all occupied tag numbers and turn them into an array
BUSY=$( 
	echo "$FULL" | 
	awk -v RS='{' '/"busy":true/ {
		beg = index($0, busy); 
		print substr($0, beg+8, 1)
	}' |
	tr '\n' ' '
)
BUSY=( $BUSY )
arr_len=${#BUSY[@]}

# Get current workspace (monitor?) index
WORKSPACE=$( 
	echo "$FULL" |
	awk -v RS='{' '/"layout".*"index".*"tags":/ { 
		beg = index($0, "index"); 
		print substr($0, beg+7, 1) 
	}'
)


# Below identifies the destination tag ($DEST), or the tag to go to
DEST=""
if [[ "$1" == "+" ]]; then
	DEST=${BUSY[0]}
	for (( i=0; i<arr_len; i++ )); do
		if (( ${BUSY[$i]} > $NOW )); then
			DEST=${BUSY[$i]}
			break
		fi
	done
elif [[ "$1" == "-" ]]; then
	DEST=${BUSY[$arr_len-1]}
	for (( i=$arr_len-1; i>=0; i-- )); do
		if (( ${BUSY[$i]} < $NOW )); then
			DEST=${BUSY[$i]}
			break
		fi
	done
else
	echo "Error: Pass \"+\" or \"-\" as parameter to run"
	exit 1
fi

# Go to tag $DEST on $WORKSPACE (haven't tested much with workspaces).
leftwm-command "SendWorkspaceToTag $WORKSPACE $(($DEST-1))"


```

**Comment by VuiMuich** [comment]
Thanks for this elaborate post.

I have thought about this functionality a couple of times myself and I believe it has been discussed loosely at some point.

I don't know, why we still missed to fix this name inconsistency, but `GoToTag` in external commands is `SendWorkspaceToTag`. Also there was a Typo at some point where it said `GotoTag`, but I don't recall at the moment if this was exposed to the user or just in the code.
However wir `SendWorkspaceToTag` you should be able to get your (very nice, I have to say) bash script enhanced as desired.

Nevertheless I agree, this would be nice to have as an internal option. 

I have thought of this in two ways:
- It could be a mode `skip_unoccupied` like `focus_behaviour` or `layout_mode`
- it could be an optional argument to the `FocusNextTag` and `FocusPreviousTag` commands

**Reply by gabberdancecat** [reply]
I like the idea of having it as an optional argument to ```FocusNextTag``` and ```FocusPreviousTag```, since I could bind a standard ```FocusNextTag``` to one keybind, and that same command but with the optional argument to another. I'll then be able to switch between both modes of switching tags easily. 

But both ways should work well.

I also updated the bash script and edited the one I posted above to match it. It's now buttery smooth and should now also work with workspaces/multi-monitors. 

This bash script is my Magnum Opus---my greatest creation thus far. It works pretty well lol
